pkg/dockerregistry/server/tagservice.go

```diff
@@ -0,0 +1,204 @@
+		return distribution.Descriptor{}, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	te := imageapi.LatestTaggedImage(imageStream, tag)
+	if te == nil {
+		return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+	}
+	dgst, err := digest.ParseDigest(te.Image)
+	if err != nil {
+		return distribution.Descriptor{}, err
+	}
+
+	if !t.repo.pullthrough {
+		image, err := t.repo.getImage(dgst)
+		if err != nil {
+			return distribution.Descriptor{}, err
+		}
+
+		if !isImageManaged(image) {
+			return distribution.Descriptor{}, distribution.ErrTagUnknown{Tag: tag}
+		}
+	}
+
+	return distribution.Descriptor{Digest: dgst}, nil
+}
+
+func (t tagService) All(ctx context.Context) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+		if !t.repo.pullthrough {
+			managed, found := managedImages[history.Items[0].Image]
+			if !found {
+				dgst, err := digest.ParseDigest(history.Items[0].Image)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+					continue
+				}
+
+				image, err := t.repo.getImage(dgst)
+				if err != nil {
+					context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
+					continue
+				}
+				managed = isImageManaged(image)
+				managedImages[history.Items[0].Image] = managed
+			}
+
+			if !managed {
+				continue
+			}
+		}
+		tags = append(tags, tag)
+	}
+	return tags, nil
+}
+
+func (t tagService) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+	tags := []string{}
+
+	imageStream, err := t.repo.getImageStream()
+	if err != nil {
+		context.GetLogger(ctx).Errorf("Error retrieving ImageStream %s/%s: %v", t.repo.namespace, t.repo.name, err)
+		return tags, distribution.ErrRepositoryUnknown{Name: t.repo.Named().Name()}
+	}
+
+	managedImages := make(map[string]bool)
+
+	for tag, history := range imageStream.Status.Tags {
+		if len(history.Items) == 0 {
+			continue
+		}
+
+		dgst, err := digest.ParseDigest(history.Items[0].Image)
+		if err != nil {
+			context.GetLogger(ctx).Errorf("bad digest %s: %v", history.Items[0].Image, err)
+			continue
+		}
+
+		if dgst == desc.Digest {
+			if !t.repo.pullthrough {
+				managed, found := managedImages[history.Items[0].Image]
+				if !found {
+					image, err := t.repo.getImage(dgst)
+					if err != nil {
+						context.GetLogger(ctx).Errorf("unable to get image %s/%s %s: %v", t.repo.namespace, t.repo.name, dgst.String(), err)
```

Comment:
I think it's wrong. We need to see why the matched digest was rejected. Otherwise we'll see it in the etcd and don't understand why it is not returned by this function.